File: frontend/types/replan.ts
/**
 * Mirrors backend/app/schemas/replan.py - keep these in sync manually
 * (same convention as types/planner.ts, types/task.ts).
 */

export interface TaskAssignment {
  task_id: string;
  week_number: number;
  day_number: number | null;
}

export interface TaskReschedule {
  task_id: string;
  title: string;
  estimated_minutes: number | null;
  previous_week: number;
  new_week: number;
}

export interface ReplanProposal {
  plan_id: string;
  reason: string;
  changes: TaskReschedule[];
  remaining_effort_minutes: number;
  // "YYYY-MM-DD" or null (a plan with nothing left has no estimate)
  new_completion_estimate: string | null;
  deadline_at_risk: boolean;
  task_assignments: TaskAssignment[];
}

export interface ReplanEvent {
  id: string;
  plan_id: string;
  reason: string;
  summary: Record<string, unknown>;
  created_at: string;
}

File: backend/app/schemas/replan.py
"""Request/response schemas for the Adaptive Planning Engine (replan)."""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class TaskAssignmentSchema(BaseModel):
    """One task's week/day placement - the same shape flows out of
    POST .../replan (inside a proposal) and back into
    POST .../replan/accept (echoed by the client), since the server
    never stores a pending proposal between the two calls - see
    app/api/v1/replan.py."""

    task_id: uuid.UUID
    week_number: int
    day_number: int | None = None


class TaskRescheduleRead(BaseModel):
    task_id: uuid.UUID
    title: str
    estimated_minutes: int | None
    previous_week: int
    new_week: int


class ReplanProposalRead(BaseModel):
    """Response for POST /planner/plans/{plan_id}/replan - a preview only,
    nothing in the database has changed yet."""

    plan_id: uuid.UUID
    reason: str
    changes: list[TaskRescheduleRead]
    remaining_effort_minutes: int
    new_completion_estimate: date | None
    deadline_at_risk: bool
    task_assignments: list[TaskAssignmentSchema]


class ReplanAcceptRequest(BaseModel):
    """The client sends back exactly what POST .../replan just proposed -
    `reason`/`changes`/etc. so the accepted event can be logged verbatim,
    and `task_assignments` so the server knows what to actually write."""

    reason: str
    changes: list[TaskRescheduleRead]
    remaining_effort_minutes: int
    new_completion_estimate: date | None
    deadline_at_risk: bool
    task_assignments: list[TaskAssignmentSchema]


class ReplanEventRead(BaseModel):
    """One row of replan history (GET /planner/plans/{plan_id}/replan/history)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plan_id: uuid.UUID
    reason: str
    summary: dict
    created_at: datetime
